### PR TITLE
Enable decompression of response within LowLevelRestClient

### PR DESCRIPTION
**Pull request by Hakky54**
This pull request is to enable the RestHighLevelClient and LowLevelRestClient to automatically handle an Elasticsearch response which has compressed content. It was already possible to do this with the low level client manually, see here [AsyncElasticsearchCompressedRequest](https://gist.github.com/Hakky54/e818a2d8d7d8153cd0ef89ca4143d805). Compression can enabled within a node configuration with the following property: `http.compression: true` Compression can be triggered by a request from a client. Therefor you also need to provide additional information within the header of the request to Elasticsearch if a client really wants to enable it. That is possible with the following RequestOptions:

    RequestOptions.Builder requestOptions = RequestOptions.DEFAULT.toBuilder()
        .addHeader("Accept-Encoding", "gzip")

With these two properties Elasticsearch will return a gzip compressed response body. The RestHighLevelClient can send a request with the above request options but it couldn't handle the response yet. This feature request contains the option to handle a compressed response if it is compressed with gzip.

Initial changes for the RestHighLevelClient have been made within this pull request: https://github.com/elastic/elasticsearch/pull/53533
Original feature requests: https://github.com/elastic/elasticsearch/issues/53555 and https://github.com/elastic/elasticsearch/issues/24349

This pull request is an improvement on the initial pull request. The LowLevelRestClient will get the responsibility to handle a compressed response and therefor this feature will also be supported within the RestHighLevelClient.

### CODE REVIEW

**Comment by Hakky54** [comment]
@andreidan have you had a chance to look at this, anything you want done differently?

File: qa/smoke-test-http/src/test/java/org/elasticsearch/http/HttpCompressionIT.java [hunk 1]
```diff
@@ -57,6 +64,30 @@ public void testUncompressedResponseByDefault() throws IOException {
         response = client().performRequest(request);
         assertEquals(201, response.getStatusLine().getStatusCode());
         assertNull(response.getHeader(HttpHeaders.CONTENT_ENCODING));
+        assertThat(response.getEntity(), is(not(instanceOf(GzipDecompressingEntity.class))));
+    }
+
+    public void testCompressesGetDocumentResponseIfRequested() throws IOException {
```
Comment (andreidan) [inline qa/smoke-test-http/src/test/java/org/elasticsearch/http/HttpCompressionIT.java:70]: shall we rename this to the existing test name `testCompressesResponseIfRequested` and delete the old version? (as this is more thorough as it's reading the body)

File: client/rest-high-level/src/test/java/org/elasticsearch/client/RestHighLevelClientTests.java [hunk 2]
```diff
@@ -326,59 +322,6 @@ public void testParseEntity() throws IOException {
         }
     }
 
-    public void testParseCompressedEntity() throws IOException {
```
Comment (andreidan) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/RestHighLevelClientTests.java:329]: these high level rest client decompression tests are still valid right?

**Review by andreidan — CHANGES_REQUESTED** [review]
Thanks for working on this @Hakky54. I've left a couple of minor suggestions/questions

File: qa/smoke-test-http/src/test/java/org/elasticsearch/http/HttpCompressionIT.java (hunk 1 above)
Comment (Hakky54) [reply qa/smoke-test-http/src/test/java/org/elasticsearch/http/HttpCompressionIT.java]: Sure, just fixed it

File: client/rest-high-level/src/test/java/org/elasticsearch/client/RestHighLevelClientTests.java (hunk 2 above)
Comment (Hakky54) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/RestHighLevelClientTests.java:329]: Well the decompression happens within the llrc when executing request and parsing the response. When the response is given to the rhlc the compressed data is already decompressed. In any kind of scenario the rhlc will always get a decompressed http enity, so therefor I can't unit test that method within the rhlc anymore.

What do you think, should we drop this unit test or do you think there is a way to validate this use case within the rhlc?

Comment (andreidan) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/RestHighLevelClientTests.java:329]: Sorry, I was a bit confused as we previously dropped the mocks in this test, but you're right, the decompression is not a `parseEntity` matter anymore. 

My aim was to make sure we test the HLRC, irrespective of who in the implementation chain performs it (it is now the LLRC's responsibility but that might change).

 After looking at this a bit closer we are covered here by the `HighLevelRestClientCompressionIT` which makes sure the HLRC is tested.

**Review by andreidan — APPROVED** [review]
LGTM, thanks for working on this @Hakky54

**Comment by andreidan** [comment]
ok to test

**Comment by Hakky54** [comment]
Awesome! Thank you @andreidan!

**Pull request merged**